# Import used packages

In [1]:
import pandas as pd
import folium
import branca.colormap as cm
from h3 import h3
from shapely.geometry import Polygon

# Import data and convert to geopandas df

In [2]:
trips = pd.read_csv('../data/processed/trips.csv')

In [3]:
trips.head(2)

bike  bike_type  identification           start_time             end_time  \
0  20507         29        10278649  2019-04-14 16:02:00  2019-04-14 16:09:00   
1  20507         29        10278649  2019-04-15 16:08:00  2019-04-15 16:15:00   

   duration_sec  start_lng  start_lat   end_lng    end_lat  start_place  \
0         420.0   8.884911  53.078923  8.884911  53.078923       2985.0   
1         420.0   8.884911  53.078923  8.884911  53.078923       2985.0   

   end_place  
0     2985.0  
1     2985.0

## Add h3 hexagon (resolution 9) addresses to trips data frame

In [4]:
resolution = 9

sHex = []
eHex = []

for index, row in trips.iterrows():
    sHex.append(h3.geo_to_h3(row['start_lat'], row['start_lng'], resolution))
    eHex.append(h3.geo_to_h3(row['end_lat'], row['end_lng'], resolution))

In [5]:
trips['sHex'] = sHex
trips['eHex'] = eHex

In [6]:
trips.head(2)

bike  bike_type  identification           start_time             end_time  \
0  20507         29        10278649  2019-04-14 16:02:00  2019-04-14 16:09:00   
1  20507         29        10278649  2019-04-15 16:08:00  2019-04-15 16:15:00   

   duration_sec  start_lng  start_lat   end_lng    end_lat  start_place  \
0         420.0   8.884911  53.078923  8.884911  53.078923       2985.0   
1         420.0   8.884911  53.078923  8.884911  53.078923       2985.0   

   end_place             sHex             eHex  
0     2985.0  891f154a56fffff  891f154a56fffff  
1     2985.0  891f154a56fffff  891f154a56fffff

In [7]:
trips['count'] = 1 # Add count column to count number of trips per hexagon

## Visualize start locations

In [8]:
starts = trips.groupby(['sHex']).agg({'count':'sum'}) # Aggregate all starts related to the corresponding hexagon

In [9]:
starts.reset_index(inplace=True)

There are many hexagons with only one trip in 2019, so we visualize only hexagons with at least 50 trips in 2019.

In [10]:
starts.sort_values(by='count')

sHex  count
1133  891f155b323ffff      1
664   891f155800bffff      1
662   891f1558003ffff      1
650   891f154b58bffff      1
647   891f154b57bffff      1
...               ...    ...
970   891f15599cfffff   2395
940   891f1559947ffff   2398
966   891f15599bbffff   3291
944   891f1559957ffff   3707
930   891f1559917ffff   9799

[1134 rows x 2 columns]

## Functions for visualization

In [11]:
# Function that takes a list of and latitude and longitude coordinates reverses them and returns a list.
def reverse_lat_lon(hex_coords):
    geom_hex = []

    for lat_lon in hex_coords:
        geom_hex.append([lat_lon[1], lat_lon[0]])

    return geom_hex

# Function that draws a polygon onto a folium map.
def add_poly_to_map(polygon, m, color, fillColor, weight, colormap):
    
    style_function = lambda x: {"weight":0, 
                                'color':'black',
                                'fillColor':colormap(weight), 
                                'fillOpacity':0.4}
    
    popup = folium.Popup('Number of trips in 2019: ' + str(weight))
    
    gj = folium.GeoJson(polygon, style_function=style_function)
    
    gj.add_child(popup)
    
    gj.add_to(m)

# Function that draws one or multiple h3 hexagon onto a folium map.
def add_hex_to_map(hex_address, m, color, fillColor, weight, colormap):
    if isinstance(hex_address, str):
        hex_poly = Polygon(reverse_lat_lon(h3.h3_to_geo_boundary(h3_address=hex_address)))

        add_poly_to_map(hex_poly, m, color, fillColor, weight, colormap)

    if isinstance(hex_address, list):

        for address in hex_address:
            hex_poly = Polygon(reverse_lat_lon(h3.h3_to_geo_boundary(h3_address=address)))

            add_poly_to_map(hex_poly, m, color, fillColor, weight, colormap)

In [12]:
# Create map
m = folium.Map(location=[53.0792962,8.8016937], zoom_start=12)

# Add style elements to map
folium.TileLayer('CartoDB positron',name="Light Map",control=False).add_to(m)
colormap = cm.LinearColormap(colors=['blue','red'], index=[50,3800],vmin=0,vmax=3800) # Create colormap
colormap.caption = "number of starts in 2019"
colormap.add_to(m) # Add colormap to map

# Add hexagons with more than 49 to map
for index, row in starts.sort_values(by='count')[:len(starts)-1][starts['count'] > 49].iterrows():
    weight = row['count']
    add_hex_to_map(row['sHex'], m, 'royalblue', 'royalblue', weight, colormap)

# Add hexagon with most trips to the map - seperately for nicer visualization
add_hex_to_map('891f1559917ffff', m, 'royalblue', 'royalblue', 9799, colormap)

# Save map to reports
m.save('../reports/figures/start_locations_hexmap_2019.html')

# Show map
m

/Users/timockenga/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:11: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  # This is added back by InteractiveShellApp.init_path()


## Visualize end locations

In [13]:
ends = trips.groupby(['eHex']).agg({'count':'sum'}) # Aggregate all ends related to the corresponding hexagon

In [14]:
ends.reset_index(inplace=True)

There are many hexagons with only one trip in 2019, so we visualize only hexagons with at least 50 trips in 2019.

In [15]:
ends.sort_values(by='count')

eHex  count
577   891f154ae37ffff      1
581   891f154ae6bffff      1
1153  891f155b323ffff      1
575   891f154ae1bffff      1
570   891f154ad4bffff      1
...               ...    ...
981   891f15599cfffff   2347
951   891f1559947ffff   2373
977   891f15599bbffff   3198
955   891f1559957ffff   3660
941   891f1559917ffff   9406

[1155 rows x 2 columns]

In [16]:
# Create map
m = folium.Map(location=[53.0792962,8.8016937], zoom_start=12) # Create map

# Add style elements to map
folium.TileLayer('CartoDB positron',name="Light Map",control=False).add_to(m)
colormap = cm.LinearColormap(colors=['blue','red'], index=[50,3800],vmin=0,vmax=3800) # Create colormap
colormap.caption = "number of starts in 2019"
colormap.add_to(m) # Add colormap to map

# Add hexagons with more than 49 to map
for index, row in ends.sort_values(by='count')[:len(ends)-1][ends['count'] >= 50].iterrows():
    weight = row['count']
    add_hex_to_map(row['eHex'], m, 'royalblue', 'royalblue', weight, colormap)

# Add hexagon with most trips to the map - seperately for nicer visualization
add_hex_to_map('891f1559917ffff', m, 'royalblue', 'royalblue', 9406, colormap)

# Save map to reports
m.save('../reports/figures/end_locations_hexmap_2019.html')

# Show map
m

/Users/timockenga/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:11: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  # This is added back by InteractiveShellApp.init_path()


As one can see start and end locations are almost the same regarding the number of trips related to them. In genereal, there are some more end locations (1155 - 1134 = 21) than start locations (for hexagons with resolution of 9) that have at least 50 related trips in 2019.